### PR TITLE
LF-3005(3): Unit tests for unit component

### PR DESCRIPTION
**Pull request by SayakaOno**
**Description**

- create a `UnitTest` class for testing the unit component (I was not sure about what the class name should be and am open to any suggestion!)
- add tests for the unit component

I created a new folder called "test-utils". If it should be camelCase, let me know. (I did the same as what I saw in an online course, but I am not sure about the convention)

How to run test:
- UI
   ```
    pnpm storybook
   ```
   http://localhost:6006/?path=/story/components-input-unit--default
   Click "Interactions" tab. (You might need to click the Rerun button.)
   Screenshot 2023-03-24 at 12 52 28 PM
- command line
   ```
   pnpm storybook
   ```
   ```
   pnpm test-storybook src/stories/Form/Unit/ 
   ```
**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/webapp/src/test-utils/storybook/unit.jsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  async testMaxValueError() {
+    const errorContainer = await this.canvas.findByTestId(`${this.testId}-errormessage`);
+    expect(errorContainer).toHaveTextContent(/Please enter a value between 0-[0-9]./);
+  }
+
+  async testNoError() {
+    await waitFor(() => {
+      const clearButton = this.canvas.queryByTestId(`${this.testId}-errorclearbutton`);
+      const requiredErrorElement = this.canvas.queryByText('Required');
+      const maxValueErrorElement = this.canvas.queryByTestId(`${this.testId}-errorclearbutton`);
```
Comment (Duncan-Brain) [inline packages/webapp/src/test-utils/storybook/unit.jsx:117]: @SayakaOno 

Should these be await this.canvas.findByTestId(`${this.testId}-errormessage`) like above? 

It seems like clearButton = maxValueErrorElement here

File: packages/webapp/src/stories/Form/Input/Unit.stories.jsx [hunk 2]
```diff
@@ -60,6 +65,39 @@ Default.args = {
   system: 'imperial',
   required: true,
 };
+// Test simple input and errors
+Default.play = async ({ canvasElement }) => {
+  const test = new UnitTest(canvasElement, 'unit');
+
+  await test.testSelectedUnit(UnitTest.getUnitLabelByValue('ft2'));
```
Comment (Duncan-Brain) [inline packages/webapp/src/stories/Form/Input/Unit.stories.jsx:72]: @SayakaOno 

Just wondering why some testSelectedUnit use UnitTest.getUnitLabelByValue() but others don't.

File: packages/webapp/src/stories/Form/Input/Unit.stories.jsx [hunk 3]
```diff
@@ -274,8 +671,316 @@ StringAsUnit.args = {
...
+};
+ImperialDefaultValueWithDecimals.play = async ({ canvasElement }) => {
+  const test = new UnitTest(canvasElement, 'unit', seedYield);
+
+  await test.testVisibleValue(test.convertDBValueToDisplayValue(0.0425, 'lb'));
+  await test.testHiddenValue(0.0425);
+  await test.testSelectedUnit('lb');
+};
+
+export const ImeprialAreaTotalAreaAutoConversion = Template.bind({});
```
Comment (Duncan-Brain) [inline packages/webapp/src/stories/Form/Input/Unit.stories.jsx:729]: small spelling errors here

**Review by Duncan-Brain — APPROVED** [review]
This looks like it was a lot of work! All the testing works for me. 

Because of the large scope I don't think I can give all of these tests a lot of attention and so I trust that you have very good reasons for each test. I tried to look through them all briefly.

My only two comments are very superficial and I am unsure about them-- feel free to ignore me:

1. Naming tests: I am not sure the convention. But for me I think re-describing the expect() statements could be good names. For example if you were in Unit.stories.jsx instead of: 
`test.testDisabledStatus()`
you would see:
`test.visibleInputAndComboxIsDisabled()`

2. Wondering what you think about the placement of Units within the storybook folder structure? 
- Should units be buried in input? or maybe reflect our regular component folder in /src
- There are lots of tests in Units is there any way to put similar tests in a folder? (ie all the unmatch)

File: packages/webapp/src/test-utils/storybook/unit.jsx (hunk 1 above)
Comment (SayakaOno) [reply packages/webapp/src/test-utils/storybook/unit.jsx:117]: There are three ways to query elements in testing-library:
- getBy... (find an existing element)
- findBy... (used with `await` to wait for an element to show up)
- queryBy... ("find" an element that does not exist)

I used `waitFor` and `queryByTestId` here to wait for errors to be gone.
Here is the reference: https://testing-library.com/docs/guide-disappearance/

Thank you finding the mistake...!

**Comment by SayakaOno** [comment]
@Duncan-Brain 
Thank you for reviewing!
1. I tried renaming tests. Please let me know if you have better ideas. I am not good at naming...
2. I moved units to the Form folder and created `SystemUnmatch` folder. I could do more cleaning, but I think it's okay for now.

File: packages/webapp/src/stories/Form/Input/Unit.stories.jsx (hunk 2 above)
Comment (Duncan-Brain) [reply packages/webapp/src/stories/Form/Input/Unit.stories.jsx]: Ohh okay thanks! haha

**Comment by Duncan-Brain** [comment]
@SayakaOno 

Naming is impossible without standards ahah. Maybe we can look up a standard for the future because I do not know what is right either. But it was hard to find an easy standard on simple google search.

The organization is nice! It definitely chunks it up! The new test command as a result for anyone seeing this : 
`pnpm test-storybook src/stories/Form/Unit/ `

File: packages/webapp/src/stories/Form/Unit/SystemUnmatch.stories.jsx [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        hookFromWatch={watch}
+        control={control}
+        {...props}
+      />
+    </form>
+  );
+};
+
+export default {
+  title: 'Form/Unit/SystemUnmatch',
```
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/webapp/src/stories/Form/Unit/SystemUnmatch.stories.jsx]: Totally agree

**Review by kathyavini — APPROVED** [review]
Sorry for the late review!

The test utilities look excellent and exhaustive. I think the new naming is very clear.

I can't quite wrap my head around the system unmatching for some reason, but I trust you on those so please do merge 🙏

**Comment by SayakaOno** [comment]
Thank you Joyce!!

**Pull request merged**